### PR TITLE
Homework 4

### PR DESCRIPTION
**Pull request by shivakcpm**
Add movie , edit movie, delete movie and sort functionality added

### CODE REVIEW

**Review by alionaSipchuk — CHANGES_REQUESTED** [review]
Too many mistakes, done poorly, look very carefully at the comments and study the material of the lecture again !!

fix all mistakes, remember that many of them apply to the whole homework!

fix if the ESLINT formatting is mandatory!

File: src/consts/movies.json [hunk 1]
```diff
@@ -2,57 +2,89 @@
...
     {
       "title": "Pulp Fiction",
       "src": "/public/pulp-fiction.PNG",
-      "year": "2004",
+      "releaseDate": "2004-10-30",
       "category": "Action & Adventure",
+      "movieURL": "movie.com",
+      "genre": "Documentory",
+      "overview": "overview",
+      "runtime": "154",
```
Comment (alionaSipchuk) [inline src/consts/movies.json:76]: why did you add this prop but never use it? let's remove it

**Review by alionaSipchuk — CHANGES_REQUESTED** [review]

File: src/components/App/App.js [hunk 2]
```diff
@@ -1,26 +1,53 @@
...
+    let [allMovies, setAllMovies] = useState(movieList);
+    const [query, setQuery] = useState('');
 
-    const onSearch = (query) => {
-        const trimmedQuery = query.trim();
+    const addMovie = (movie) => {
+        movie.id = allMovies.length;
+        allMovies.push(movie);
+        setAllMovies(allMovies);
+        onSearch(query);
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:16]: if added movie match with the current search criteria it should shown without searching again.

File: src/components/App/App.js [hunk 3]
```diff
@@ -1,26 +1,53 @@
...
+            ? allMovies.filter((item) =>
                 new RegExp(trimmedQuery, 'i').test(item.title)
             )
-            : movieList;
+            : [...allMovies];
+        setQuery(trimmedQuery);
         setMovies(filteredMovies);
     };
 
+    const  [movies, setMovies] =  useState(allMovies);
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:42]: here movies are after applying the search query

File: src/consts/movies.json (hunk 1 above)
Comment (shivakcpm) [reply src/consts/movies.json:76]: we are using this in add movie functionality

File: src/components/AddMovie/AddMovieComponent.js [hunk 4]
```diff
@@ -0,0 +1,164 @@
...
+          genre: '',
+          overview: '',
+          runtime: '',
+          src:'/public/avengers.PNG',
+          category:'Action & Adventure'
+      };
+
+      const state =  {
+          movie: this.props.editMode
+              ? { ...this.props.movie }
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/AddMovie/AddMovieComponent.js:41]: here i am updating the state to intial state when user resets the form. in edit mode initial state will be the movie passed as prop.

File: src/components/AddMovie/AddMovieComponent.js [hunk 5]
```diff
@@ -0,0 +1,164 @@
...
+              : movie,
+          showErrors:false
+      };
+      return state;
+  }
+
+  handleReset = () => {
+      this.setState(this.getInitialState());
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/AddMovie/AddMovieComponent.js]: I am using get initial state function in two places , in constructor and in  reset methods.

File: src/components/App/App.js [hunk 6]
```diff
@@ -1,26 +1,53 @@
...
 export default function App() {
-    const [movies, setMovies] = useState(movieList);
+    // eslint-disable-next-line prefer-const
+    let [allMovies, setAllMovies] = useState(movieList);
+    const [query, setQuery] = useState('');
 
-    const onSearch = (query) => {
-        const trimmedQuery = query.trim();
+    const addMovie = (movie) => {
+        movie.id = allMovies.length;
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js]: i do not have access to the all movies in addmovie component to generate id, IId will be autogenerated  if movies apis are integrated.

File: src/components/App/App.js [hunk 7]
```diff
@@ -1,26 +1,53 @@
...
     };
 
+    const  [movies, setMovies] =  useState(allMovies);
+
     return (
         <React.StrictMode>
             <div className="container">
-                <HeaderComponent onSearch={onSearch} />
-                <ContentHolderComponent movies={movies} />
+                <myContext.Provider value={{ addMovie, editMovie, onSearch, deleteMovie, movies }}>
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js]: I renamed myContext to Context and passing editMovie and deleteMovie methods.

File: src/components/MovieCard/MovieCardComponent.js [hunk 8]
```diff
@@ -1,21 +1,95 @@
...
+        }
+    }
+
+    toggleDeleteDialog = () => {
+        this.toggleBodyScroll(this.state.openDeleteDialog);
+        this.setState({ openDeleteDialog:!this.state.openDeleteDialog });
+    }
+
+    deleteMovie = () => {
+        this.toggleDeleteDialog();
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/MovieCard/MovieCardComponent.js]: toggle delete dialog method i am using to show or hide delete dialog. here once user confirms delete i am hiding the dialog.

File: src/components/MovieCard/MovieCardComponent.js [hunk 9]
```diff
@@ -1,35 +1,99 @@
...
+
+  toggleEditMovieDialog = () => {
+    this.setState({ isOpenEditDialog: !this.state.isOpenEditDialog });
+  };
+
+  toggleDeleteDialog = () => {
+    this.setState({ isOpenDeleteDialog: !this.state.isOpenDeleteDialog });
+  };
+
+  deleteMovie = () => {
```
Comment (alionaSipchuk) [inline src/components/MovieCard/MovieCardComponent.js:46]: rename method as onDeleteMehod

File: src/components/MovieCard/MovieCardComponent.js [hunk 10]
```diff
@@ -1,35 +1,99 @@
...
+        <img src={src} alt={title}></img>
+        <div className="menu-icon" onClick={() => this.menuToggler(true)}>
+          &#xFE19; {this.contextType}{' '}
+        </div>
+        {this.state.isOpenContext && (
+          <ContextMenu
+            menu={MENU}
+            onMenuItemClicked={this.onMenuItemClicked}
+            closeHandler={this.menuToggler}
+          ></ContextMenu>
```
Comment (alionaSipchuk) [inline src/components/MovieCard/MovieCardComponent.js:67]: no need to write two tags without children!!!!!!!!

```
<ContextMenu
   menu={MENU}
   onMenuItemClicked={this.onMenuItemClicked}
   closeHandler={this.menuToggler}
/>
```

File: src/components/MovieCard/MovieCardComponent.js [hunk 11]
```diff
@@ -1,35 +1,99 @@
...
+  };
+
+  deleteMovie = () => {
+    this.props.deleteMovie(this.props.movie);
+    this.toggleDeleteDialog();
+  };
+
+  render() {
+    const {
+      movie: { title, releaseDate, category, src }
```
Comment (alionaSipchuk) [inline src/components/MovieCard/MovieCardComponent.js:53]: add also get movie from props here!!

```
const {
      movie: { title, releaseDate, category, src },
     movie,
} = this.props;
```

File: src/components/MovieList/MovieListComponent.js [hunk 12]
```diff
@@ -4,25 +4,25 @@ import PropTypes from 'prop-types';
...
+    if (movies.length === 0) {
+      throw new Error('no Movies found');
     }
+
+    return (
+      <>
+        <div className="movies-count">{movies.length} movies found</div>
+        <div className="movie-list">
+          {movies.map(value => {
+            return <MovieCardComponent movie={value} key={value.id} editMovie={this.props.editMovie} deleteMovie={this.props.deleteMovie}/>;
```
Comment (alionaSipchuk) [inline src/components/MovieList/MovieListComponent.js:18]: no need to write return and wrap into curly braces!

```
{movies.map(value => (
   <MovieCardComponent
       movie={value}
       key={value.id}
       editMovie={this.props.editMovie}
       deleteMovie={this.props.deleteMovie}
    />
 ))}
```

**Review by alionaSipchuk — CHANGES_REQUESTED** [review]
Pay attention! many mistakes are repeated from hometask to hometask!

if in the comment you will see a lot of !!! - in mean that i was very ANGRY  😊 😊 😊 

AS SOON AS POSSIBLE FOX COMMENTS!

File: src/components/App/App.js [hunk 13]
```diff
@@ -1,27 +1,37 @@
...
-        const trimmedQuery = query.trim();
-        const filteredMovies = trimmedQuery
-            ? movieList.filter((item) =>
-                new RegExp(trimmedQuery, 'i').test(item.title)
-            )
-            : movieList;
-        setMovies(filteredMovies);
-    };
+  const addMovie = movie => {
+    allMovies.push(movie);
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:12]: here all movies is not a state variable. movies is a state variable. its contains all the movies if there is no search query. if there is any search query , value will be the results of the search

File: src/components/App/App.js [hunk 14]
```diff
@@ -1,27 +1,37 @@
...
-    return (
-        <React.StrictMode>
-            <div className="container">
-                <HeaderComponent onSearch={onSearch} />
-                <ContentHolderComponent movies={movies} />
-            </div>
-        </React.StrictMode>
-    );
+  const setAllMovies = (newMovies) => {
+      allMovies = newMovies;
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:16]: here all movies is not a state variable. movies is a state variable. its contains all the movies if there is no search query. if there is any search query , value will be the results of the search

File: src/components/App/App.js [hunk 15]
```diff
@@ -1,27 +1,37 @@
...
+      : allMovies;
+    setMovies(filteredMovies);
+  };
+
+  return (
+    <React.StrictMode>
+      <div className="container">
+        <HeaderComponent addMovie={addMovie} onSearch={onSearch} />
+        <Context.Provider value={{ movies }}>
+          <ContentHolderComponent allMovies={allMovies} setAllMovies= {setAllMovies} />
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:32]: here all movies is not a state variable. movies is a state variable. its contains all the movies if there is no search query. if there is any search query , value will be the results of the search

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 16]
```diff
@@ -1,42 +1,70 @@
...
-            <div className="footer-bar">
-                <strong>netflix</strong>
-                <span>roulette</span>
-            </div>
-        </>
-    );
-}
+  const movies = useContext(Context).movies;
+  const [sortOrder, setSortOrder] = useState(DEFAULT_SORT_BY);
+  let allMovies = props.allMovies;
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/ContentHolder/ContentHolderComponent.js:13]: i am using allMovies variable in one or more places. it would be easy to use if i create a variable here

File: src/components/ContentHolder/ContentHolderComponent.js [hunk 17]
```diff
@@ -1,42 +1,70 @@
...
+
+  const editMovie = movie => {
+    const moviesFiltered = allMovies.filter(item => item.id !== movie.id);
+    allMovies = [...moviesFiltered, movie];
+    props.setAllMovies(allMovies);
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/ContentHolder/ContentHolderComponent.js:18]: I do not understand this logic , allMovies is an array here. you are trying to get the object by index  from array, but it will not work here.

File: src/components/App/App.js (hunk 15 above)
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:32]: movies are filtered once, if search term is empty both are same

**Review by alionaSipchuk — APPROVED** [review]
Cool, I'm very glad to reviewing this code
pls fix not critical issues))

File: src/components/ErrorBoundary/ErrorBoundary.js [hunk 18]
```diff
@@ -2,33 +2,32 @@ import React, { Component } from 'react';
...
-        return { hasError: true, errorMessage: error.errorMessage };
-    }
+  static getDerivedStateFromError(error) {
+    return { hasError: true, errorMessage: error.errorMessage };
+  }
 
-    componentDidCatch(error, errorInfo) {
-        const { logErrors = true } = this.props;
+  componentDidCatch(error, errorInfo) {
+    const { logErrors = true } = this.props;
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/ErrorBoundary/ErrorBoundary.js:15]: in homework-3 , you suggested me to use logErrors true.  I thought it is easy to debug errors if we log, any way now i will change it to false.

File: src/components/App/App.js [hunk 19]
```diff
@@ -1,27 +1,34 @@
...
+
+  const filteredMovies = useCallback(() => {
+    return movies.filter(item => new RegExp(query, 'i').test(item.title));
+  }, [movies, query]);
+
+  return (
+    <React.StrictMode>
+      <div className="container">
+        <HeaderComponent addMovie={addMovie} onSearch={onSearch} />
+        <Context.Provider value={{ movies:filteredMovies() }}>
```
[the remark this answers is not in the document]
Comment (shivakcpm) [reply src/components/App/App.js:19]: my lint config has "object-curly-spacing": [2, "always"],
